Lib/contextlib.py

```diff
@@ -152,13 +164,20 @@ def __exit__(self, typ, value, traceback):
                 try:
                     raise RuntimeError("generator didn't stop")
                 finally:
+                    del self.exc_context
                     self.gen.close()
         else:
             if value is None:
                 # Need to force instantiation so we can reliably
                 # tell if we get the same exception back
                 value = typ()
             try:
+                # If the generator handles the exception thrown into it, the
+                # exception context will revert to the actual current exception
+                # context here. In order to make the context manager behave
+                # like a normal function we set the current exception context
+                # to what it was during the context manager's __enter__
+                _set_current_exception(self.exc_context)
```

Comment:
If we were to always set `self.exc_context` back to `None`, instead of deleting it, this line could stay as it is, right?

Reply:
If we were to do the clearing at the top of the method, the adjustment here would be minor:

```suggestion
                # If the generator handles the exception thrown into it, the
                # exception context will revert to the actual current exception
                # context here. In order to make the context manager behave
                # like a normal function we set the current exception context
                # to what it was during the context manager's __enter__
                _set_current_exception(exc_context)
```